transport.go

```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
 	for {
 		select {
 		case req := <-t.requests:
-			// BUG(stevvooe): This is an awful tag allocation procedure.
-			// Replace this with something that let's us allocate tags and
-			// associate data with them, returning to them to a pool when
-			// complete. Such a system would provide a lot of information
-			// about outstanding requests.
-			tags++
-			fcall := newFcall(tags, req.message)
+			var selected Tag
+			if req.message.Type() == Tversion {
+				// Tversion can only use NOTAG.
+				if _, exists := outstanding[NOTAG]; exists {
+					req.err <- errors.New("NOTAG already in use")
+				}
+				selected = NOTAG
+			} else {
+				// The tag pool is depleted if all 65536 tags are taken, or if
+				// 65535 tags are taken and NOTAG is available.
+				if len(outstanding) > 0xFFFF {
+					req.err <- errors.New("tag pool depleted")
+				} else if len(outstanding) == 0xFFFF {
+					if _, exists := outstanding[NOTAG]; !exists {
+						req.err <- errors.New("tag pool depleted")
+					}
+				}
+
+				exists := true
+				for selected = 0; exists && selected < NOTAG; selected++ {
```

Comment:
I have reintroduced the previous selection hint. I used it in my own implementation as well, but removed it as a simplification, thinking it could somewhat be considered a premature optimisation.

Having 65535 pending requests for a single client is unrealistic. A more realistic number of alive tags would be, idk, 10 to 100 tags, in which case we get <=1KB with `λ = 1`, but still 512KB for an array. Remember that tags are client local.

I agree that iterating on a map is a terrible idea, but I was weighing it against a permanent array that would be orders of magnitude larger. Even with thousand tags open, the overhead of using a permanent array is quite significant.